src/components/DomEvents.js

```diff
@@ -127,18 +132,20 @@ function DomEvents() {
     [setEventCount],
   );
 
-  const setPreviewRef = useCallback((node) => {
-    previewRef.current = node;
-
-    if (!node) return;
-
-    addLoggingEvents(node, (event) => {
+  useEffect(() => {
+    if (!previewRef.current) return;
+    const eventListeners = addLoggingEvents(previewRef.current, (event) => {
       // insert at index 0
       event.id = buffer.current.length;
       buffer.current.splice(0, 0, event);
       setTimeout(flush, 0);
     });
-  }, []);
+    return () => {
+      eventListeners.forEach((event) =>
+        previewRef.current.removeEventListener(event.name, event.listener),
+      );
+    };
+  }, [previewRef.current]);
```

Comment:
Will this work? In my experience, `ref.current` values don't (always) trigger `effects`. As the value is mutated. That's why I used the callback method.

So, I think the `removeEventListener` would still be possible, but the `setPreviewRef` callback should get an extra condition:

```js
if (node !== previewRef.current) {
  return;
}

if (previewRef.current) {
  // removeLoggingEvents(previewRef.current, ...
}

if (node) {
  // addLoggingEvents(node, ...
}
```

Reply:
Source for the above:

https://reactjs.org/docs/hooks-reference.html#useref

> Keep in mind that `useRef` _doesn’t_ notify you when its content changes. Mutating the `.current` property doesn’t cause a re-render. If you want to run some code when React attaches or detaches a ref to a DOM node, you may want to use a callback ref instead.

Reply:
I didn't know this thing but looking around you're right. https://github.com/facebook/react/issues/16121

```diff
diff --git a/src/components/DomEvents.js b/src/components/DomEvents.js
index 7277e8c..47b3660 100644
--- a/src/components/DomEvents.js
+++ b/src/components/DomEvents.js
@@ -119,6 +119,7 @@ function DomEvents() {
   const listRef = useRef();
 
   const [eventCount, setEventCount] = useState(0);
+  const [eventListeners, setEventListeners] = useState([]);
 
   const reset = () => {
     buffer.current = [];
@@ -132,20 +133,23 @@ function DomEvents() {
     [setEventCount],
   );
 
-  useEffect(() => {
-    if (!previewRef.current) return;
-    const eventListeners = addLoggingEvents(previewRef.current, (event) => {
-      // insert at index 0
-      event.id = buffer.current.length;
-      buffer.current.splice(0, 0, event);
-      setTimeout(flush, 0);
-    });
-    return () => {
+  const setPreviewRef = useCallback((node) => {
+    if (node) {
+      previewRef.current = node;
+      const eventListeners = addLoggingEvents(node, (event) => {
+        // insert at index 0
+        event.id = buffer.current.length;
+        buffer.current.splice(0, 0, event);
+        setTimeout(flush, 0);
+      });
+      setEventListeners(eventListeners);
+    } else if (previewRef.current) {
       eventListeners.forEach((event) =>
         previewRef.current.removeEventListener(event.name, event.listener),
       );
-    };
-  }, [previewRef.current]);
+      previewRef.current = null;
+    }
+  }, []);
 
   return (
     <div className="flex flex-col h-auto md:h-full w-full">
@@ -154,7 +158,7 @@ function DomEvents() {
           <MarkupEditor markup={markup} dispatch={dispatch} />
         </div>
 
-        <div className="flex-auto h-56 md:h-full" ref={previewRef}>
+        <div className="flex-auto h-56 md:h-full" ref={setPreviewRef}>
           <Preview
             markup={markup}
             elements={result.elements}
```

Reply:
I have committed following the diff. If you have any doubt let me know 👯